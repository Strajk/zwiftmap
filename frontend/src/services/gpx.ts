import { XMLBuilder } from "fast-xml-parser";
import { LatLngAlt } from "../types";

const builder = new XMLBuilder({
  ignoreAttributes: false,
  format: true,
  suppressEmptyNode: true,
});

export function createGPX(url: string, stream: LatLngAlt[]): string {
  const content = builder.build({
    gpx: {
      "@_creator": "ZwiftMap",
      "@_xmlns": "http://www.topografix.com/GPX/1/1",
      "@_xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
      "@_xsi:schemaLocation":
        "http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd",
      "@_version": "1.1",
      metadata: {
        name: "Custom Route",
        desc: "Custom Route on Zwift",
        author: {
          name: "ZwiftMap",
        },
        link: {
          "@_href": url,
        },
      },
      trk: {
        name: "Custom Route",
        link: {
          "@_href": url,
        },
        type: "Ride",
        trkseg: stream.map(([lat, lng, alt]) => ({
          trkpt: {
            "@_lat": lat,
            "@_lon": lng,
            ele: alt,
          },
        })),
      },
    },
  });
  return `<?xml version="1.0" encoding="UTF-8"?>\n${content}`;
}
