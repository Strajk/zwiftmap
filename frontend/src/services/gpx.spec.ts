import { createGPX } from "./gpx";
import { expect, it } from "vitest";

it("returns a gpx file", async () => {
  const string = createGPX("https://zwiftmap.com", [
    [10, 20, 30],
    [11, 21, 31],
    [12, 22, 32],
  ]);
  expect(string).toMatchSnapshot();
});
