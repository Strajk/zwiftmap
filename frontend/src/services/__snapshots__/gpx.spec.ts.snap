// Vitest Snapshot v1

exports[`returns a gpx file 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\"?>
<gpx creator=\\"ZwiftMap\\" xmlns=\\"http://www.topografix.com/GPX/1/1\\" xmlns:xsi=\\"http://www.w3.org/2001/XMLSchema-instance\\" xsi:schemaLocation=\\"http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd\\" version=\\"1.1\\">
  <metadata>
    <name>Custom Route</name>
    <desc>Custom Route on Zwift</desc>
    <author>
      <name>ZwiftMap</name>
    </author>
    <link href=\\"https://zwiftmap.com\\"/>
  </metadata>
  <trk>
    <name>Custom Route</name>
    <link href=\\"https://zwiftmap.com\\"/>
    <type>Ride</type>
    <trkseg>
      <trkpt lat=\\"10\\" lon=\\"20\\">
        <ele>30</ele>
      </trkpt>
    </trkseg>
    <trkseg>
      <trkpt lat=\\"11\\" lon=\\"21\\">
        <ele>31</ele>
      </trkpt>
    </trkseg>
    <trkseg>
      <trkpt lat=\\"12\\" lon=\\"22\\">
        <ele>32</ele>
      </trkpt>
    </trkseg>
  </trk>
</gpx>
"
`;
